File: validation/cases/butterworth_filter/butterworth_filter.m
% butterworth_filter.m -- 4th-order Butterworth lowpass filter step response.
% Cascaded biquad sections matching ctrlpp::make_butterworth<4>().
%
% Usage: octave --no-gui butterworth_filter.m > butterworth_filter_octave.csv

pkg load signal;

% Design 4th-order Butterworth at 50 Hz, sampled at 1000 Hz
[b, a] = butter(4, 50 / 500);  % normalized cutoff = fc / (fs/2)

% Step response
n_samples = 200;
x = ones(1, n_samples);
y = filter(b, a, x);

fprintf('sample,response\n');
for i = 1:n_samples
    fprintf('%.15e,%.15e\n', i-1, y(i));
end

File: validation/cases/butterworth_filter/butterworth_filter.cpp
// butterworth_filter.cpp -- 4th-order Butterworth matching butterworth_filter.m
// Usage: ./butterworth_filter > butterworth_filter_cpp.csv

#include "ctrlpp/dsp/biquad.h"

#include <cstdio>

int main()
{
    auto filt = ctrlpp::make_butterworth<4>(50.0, 1000.0);

    constexpr int n_samples = 200;

    std::printf("sample,response\n");

    for(int i = 0; i < n_samples; ++i)
    {
        double output = filt.process(1.0);
        std::printf("%.15e,%.15e\n", static_cast<double>(i), output);
    }
}
